File: 2022-23/contest2/prog/intermediate/binary_counting.java
import java.io.*;
import java.util.*;

public class binary_counting {

	public static String reverse(String s) {
		return new StringBuilder(s).reverse().toString();
	}

	public static String to_string(int num, int base, int width) {
		if(num == 0)
			return new String("0");
		String res = new String();
		int i = 0;
		while(num > 0 || width != Integer.MAX_VALUE) {
			int d = num % base;
			res += (char)(d + '0');
			num /= base;
			if(++i >= width)
				break;
		}
		return reverse(res);
	}

	public static String remove_left(String src, String sub) {
		int pos = src.indexOf(sub);
		if(pos < 0)
			return src;
		String res = new String();
		for(int i = 0; i < src.length(); ++i) {
			if(i < pos || i >= pos + sub.length())
				res += src.charAt(i);
		}
		return res;
	}

	public static String remove_right(String src, String sub) {
		int pos = src.lastIndexOf(sub);
		if(pos < 0)
			return src;
		String res = new String();
		for(int i = 0; i < src.length(); ++i) {
			if(i < pos || i >= pos + sub.length())
				res += src.charAt(i);
		}
		return res;
	}

	public static void main(String[] args) {
		Scanner input = new Scanner(System.in);
		for(int k = 0; k < 5; ++k) {
			String line = input.nextLine();
			String bs = new String();
			for(int i = 0; i < line.length(); ++i) {
				char ch = line.charAt(i);
				String cs = to_string((int)ch, 2, 8);
				bs += cs;
			}
			int i = 0;
			for(;;++i) {
				String s = to_string(i, 2, Integer.MAX_VALUE);
				String l = remove_left(bs, s);
				String r = remove_right(l, s);
				if(bs == r)
					break;
				bs = r;
			}
			System.out.println(i-1);
		}
	}

}

File: 2022-23/contest2/prog/intermediate/binary_counting.py
import os
import sys

INT_MAX = 65536

def to_string(num, base, width = INT_MAX):
	if num == 0:
		return '0'
	res = ''
	i = 0
	while num or width != INT_MAX:
		d = num % base
		res += str(d)
		num = num // base
		i += 1
		if i >= width:
			break
	return res[::-1] # Reverse a string efficiently in Python

def remove_left(src, sub):
	pos = src.find(sub)
	if pos < 0:
		return src
	res = ''
	for i in range(len(src)):
		if i < pos or i >= (pos + len(sub)):
			res += src[i]
	return res

def remove_right(src, sub):
	pos = src.rfind(sub)
	if pos < 0:
		return src
	res = ''
	for i in range(len(src)):
		if i < pos or i >= (pos + len(sub)):
			res += src[i]
	return res

for k in range(5):
	line = input()
	bs = ''
	for ch in line:
		cs = to_string(ord(ch), 2, 8)
		bs += cs
	i = 0
	while(True):
		s = to_string(i, 2)
		l = remove_left(bs, s)
		r = remove_right(l, s)
		if bs == r:
			break
		bs = r
		i += 1
	print(i - 1)
